04-onion-routing.md

```diff
@@ -367,6 +368,102 @@ otherwise meets the amount criterion (eg. some other failure, or
 invoice timeout), however if it were to fulfill only some of them,
 intermediary nodes could simply claim the remaining ones.
 
+### Onion Messages
+
+Onion messages have an onion with an alternate `hop_payload`
+format: a `bigsize` followed by a `onionmsg_payload`.  Note that there
+is no legacy format, thus a `bigsize` of 0 means no payload.
+
+An intermediate node expects an `enctlv` which it can decrypt using
+the `blinding` which it is handed along with the onion message.
+
+1. `tlv_stream`: `onionmsg_payload`
+2. types:
+    1. type: 8 (`reply_path`)
+    2. data:
+        * [`point`:`first_node_id`]
+        * [`point`:`blinding`]
+        * [`...*onionmsg_path`:`path`]
+    1. type: 10 (`enctlv`)
+    2. data:
+        * [`...*byte`:`enctlv`]
+
+1. `tlv_stream`: `encmsg_tlvs`
+2. types:
+    1. type: 1 (`padding`)
+    2. data:
+        * [`...*byte`:`pad`]
+    1. type: 4 (`next_node_id`)
+    2. data:
+        * [`point`:`node_id`]
+    1. type: 12 (`next_blinding`)
+    2. data:
+        * [`point`:`blinding`]
+
+1. subtype: `onionmsg_path`
+2. data:
+    * [`point`:`node_id`]
+    * [`u16`:`enclen`]
+    * [`enclen*byte`:`enctlv`]
+
+#### Requirements
+
+The writer:
+- For the non-final nodes' `onionmsg_payload`:
+  - MUST set `enctlv` to a valid `encmsg_tlvs` containing `next_node_id`.
+  - MUST encrypt `enctlv` as detailed in [Route Blinding][route-blinding].
+    `ChaChaPoly-1305` encryption using an all-zero nonce).
```

Comment:
Its really kinda awkward to use ChaCha+HMAC-SHA256 for the outer onion and ChaCha+Poly1305 for the inner one. I guess you decided that saving the 16 bytes per hop is that important? But if so, why not just also use ChaChaPoly for the outer onion as well?

Reply:
Again, this is fixed already unless we want to rewrite all onion code.  The inner one is better chachapoly1305, because that we can control!

Reply:
Heh, on our end almost everything is abstracted so that the decrypt is 5 lines of code, but I can still trivially replace the MAC call cause which method is called is not abstracted. I get that your code isn't that way, but man it makes for a really awkward spec. I guess there's no risk for the creator to do any collisions and saving 16 bytes is worth it, but it may be worth it on your end to see if you can tweak things a bit to make reuse 95% of the onion decrypt code without being boxed into ChaCha-SHA256.

Reply:
As discussed during the last spec meeting, roasbeef provided a pretty good answer: crypto libraries usually don't let you use poly1305 without chacha20, which is why hmac-sha256 was initially chosen.